source/common/network/buffered_io_socket_handle_impl.h

```diff
@@ -0,0 +1,163 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/api/io_error.h"
+#include "envoy/api/os_sys_calls.h"
+#include "envoy/common/platform.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/network/io_handle.h"
+
+#include "common/buffer/watermark_buffer.h"
+#include "common/common/logger.h"
+#include "common/event/file_event_impl.h"
+#include "common/network/io_socket_error_impl.h"
+#include "common/network/peer_buffer.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * IoHandle implementation which provides a buffer as data source. It is designed to used by
+ * Network::ConnectionImpl. Some known limitations include
+ * 1. It doesn't not include a file descriptor. Do not use "fdDoNotUse".
+ * 2. It doesn't suppose socket options. Wrap this in ConnectionSocket and implement the socket
+ * getter/setter options.
+ * 3. It doesn't suppose UDP interface.
+ * 4. The peer BufferedIoSocket must be scheduled by the same scheduler to avoid data race.
+ */
+class BufferedIoSocketHandleImpl : public IoHandle,
+                                   public WritablePeer,
+                                   public ReadableSource,
+                                   protected Logger::Loggable<Logger::Id::io> {
+public:
+  BufferedIoSocketHandleImpl()
+      : owned_buffer_(
+            [this]() -> void {
+              over_high_watermark_ = false;
+              if (writable_peer_) {
+                ENVOY_LOG(debug, "Socket {} switches to low watermark. Notify {}.",
+                          static_cast<void*>(this), static_cast<void*>(writable_peer_));
+                writable_peer_->onPeerBufferWritable();
+              }
+            },
+            [this]() -> void {
+              over_high_watermark_ = true;
+              // Low to high is checked by peer after peer writes data.
+            },
+            []() -> void {}) {}
+
+  ~BufferedIoSocketHandleImpl() override { ASSERT(closed_); }
+
+  // IoHandle
+  os_fd_t fdDoNotUse() const override { return INVALID_SOCKET; }
+
+  Api::IoCallUint64Result close() override;
```

Comment:
you can remove the whitespace between all of these

Reply:
done